### PR TITLE
Fix dirty wasted seperable

### PR DESCRIPTION
**Pull request by HandyKurniawan**

### CODE REVIEW

File: quconot/implementations/mct_no_auxiliary_relative.py [hunk 1]
```diff
@@ -1,79 +1,79 @@
...
+
+from copy import deepcopy
+from typing import List
+
+from qiskit import QuantumCircuit, transpile
+
+from .mct_base import MCTBase
+
+
+class MCTNoAuxiliaryRelative(MCTBase):
```
Comment (adamglos92) [inline quconot/implementations/mct_no_auxiliary_relative.py:16]: the name should relate to qiskit, or original paper if applicable

File: quconot/implementations/mct_no_auxiliary_relative.py [hunk 2]
```diff
@@ -1,79 +1,79 @@
...
+        clean_acilla: bool = True,
+        wasted_auxiliary: bool = False,
+        separable_wasted_auxiliary: bool = False,
+    ) -> List["MCTBase"]:
+        """Generate all possible MCT implementation satisfying the requirements
+
+        relative_phase: true / false (D)
+        clean_auxiliary: true (D) / false
+        wasted_auxiliary: true / false (D)
+        separable_wasted_auxiliary: true / false (D)    # requires wasted_auxiliary set to True
```
Comment (adamglos92) [inline quconot/implementations/mct_no_auxiliary_relative.py:40]: no `#` in docstrings

File: quconot/implementations/mct_no_auxiliary_relative.py [hunk 3]
```diff
@@ -1,79 +1,79 @@
...
+        :return: a quantum circuit
+        :rtype: QuantumCircuit
+        """
+        qc = QuantumCircuit(self._n + 1)
+        if self._n == 2:
+            qc.rccx(0, 1, 2)
+        elif self._n == 3:
+            qc.rcccx(0, 1, 2, 3)
+        else:
+            raise ValueError("At the moment we cannot handle controls bigger than 3.")
```
Comment (adamglos92) [inline quconot/implementations/mct_no_auxiliary_relative.py:59]: I would say this should be handled at `__init__` only

File: quconot/implementations/mct_no_auxiliary_relative.py [hunk 4]
```diff
@@ -1,79 +1,79 @@
...
+    def num_auxiliary_qubits(self):
+        """Return number of auxiliary qubits
+
+        :return: number of auxiliary qubits
+        :rtype: int
+        """
+        return 0
+
+
+if __name__ == "__main__":
```
Comment (adamglos92) [inline quconot/implementations/mct_no_auxiliary_relative.py:75]: remove in final pull request

File: quconot/implementations/mct_recursion.py [hunk 5]
```diff
@@ -84,4 +84,6 @@ def num_auxiliary_qubits(self):
 
 
 if __name__ == "__main__":
```
Comment (adamglos92) [inline quconot/implementations/mct_recursion.py:86]: remove in final pull request

**Review by adamglos92 — CHANGES_REQUESTED** [review]
what is result.txt?

File: setup.cfg [hunk 6]
```diff
@@ -1,44 +1,44 @@
...
-exclude = test
-
-[coverage:run]
-relative_files = True
+[metadata]
+name = QuCoNot
+description = QuCoNot is a Python library for all your MCT needs.
+long_description = file: README.md
+long_description_content_type = text/markdown
+author = Adam Glos
```
Comment (adamglos92) [inline setup.cfg:6]: I don't feel OK having only my name here. Let's add all main contributors names here.

**Review by adamglos92 — COMMENTED** [review]
I didn't check the test functions, I'm going to assume they are correct based on their compliance with the MCT implementations

**Comment by HandyKurniawan** [comment]
> what is result.txt?

This is supposed to be the csv file that will be produced to check the graph dependencies.

File: setup.cfg (hunk 6 above)
Comment (HandyKurniawan) [reply setup.cfg:6]: @adamglos92 @ozlemsalehi @ankit27kh @shraddha-aangiras @HandyKurniawan 
So, should I add these names based in alphabetical order, or I put the supervisors' names last?

**Comment by HandyKurniawan** [comment]
> I didn't check the test functions, I'm going to assume they are correct based on their compliance with the MCT implementations

I tried to implement it based on the `MCT_Testing.pdf`. It is not the best implementation yet, i.e., still need optimization for the code.

File: quconot/implementations/mct_recursion.py (hunk 5 above)
Comment (HandyKurniawan) [reply quconot/implementations/mct_recursion.py:86]: noted, i will remove all the main

File: quconot/implementations/mct_no_auxiliary_relative.py (hunk 1 above)
Comment (HandyKurniawan) [reply quconot/implementations/mct_no_auxiliary_relative.py:16]: From qiskit i am using the rccx and rcccx for this implementation.

File: quconot/implementations/mct_recursion.py (hunk 5 above)
Comment (HandyKurniawan) [reply quconot/implementations/mct_recursion.py]: I have pushed the changes with the last commit

File: quconot/implementations/mct_no_auxiliary_relative.py (hunk 2 above)
Comment (HandyKurniawan) [reply quconot/implementations/mct_no_auxiliary_relative.py:40]: changed to `separable_wasted_auxiliary: true / false (D), requires wasted_auxiliary set to True`

File: setup.cfg (hunk 6 above)
Comment (HandyKurniawan) [reply setup.cfg:6]: this `author = Adam Glos, Özlem Salehi, Ankit Khandelwal, Handy Kurniawan, Shraddha Aangiras`
or
`author = Ankit Khandelwal, Handy Kurniawan, Shraddha Aangiras, Adam Glos, Özlem Salehi`?

File: quconot/implementations/mct_no_auxiliary_relative.py (hunk 1 above)
Comment (HandyKurniawan) [reply quconot/implementations/mct_no_auxiliary_relative.py:16]: image

For `rccx`

Comment (HandyKurniawan) [reply quconot/implementations/mct_no_auxiliary_relative.py:16]: image

For `rcccx`

Comment (HandyKurniawan) [reply quconot/implementations/mct_no_auxiliary_relative.py:16]: Based on the above paper, there are also more implementations:

image

Comment (HandyKurniawan) [reply quconot/implementations/mct_no_auxiliary_relative.py:16]: how about these names:
- `mct_qiskit_no_ancilla`
- `mct_qiskit_rccx`
- `mct_qiskit_rcccx`
- `mct_qiskit_recurssion`
- `mct_qiskit_vchain`
- `mct_qiskit_vchain_dirty`

Comment (adamglos92) [reply quconot/implementations/mct_no_auxiliary_relative.py:16]: `MaslovSmallCircuits`?
@ankit27kh perhaps you could help with naming?

File: setup.cfg (hunk 6 above)
Comment (adamglos92) [reply setup.cfg:6]: Maybe let's use alphabetical order in here

Comment (HandyKurniawan) [reply setup.cfg:6]: `author = Adam Glos, Ankit Khandelwal, Handy Kurniawan, Shraddha Aangiras, Özlem Salehi`

noted :)

File: quconot/implementations/mct_no_auxiliary_relative.py (hunk 1 above)
Comment (ankit27kh) [reply quconot/implementations/mct_no_auxiliary_relative.py:16]: @adamglos92 @HandyKurniawan 
changing the names of the implementations is something I have thought of earlier. I would want all implementations to have new names.
I suggest we open an issue to consider the naming issue and not hold this PR back.

File: setup.cfg (hunk 6 above)
Comment (ankit27kh) [reply setup.cfg:6]: @HandyKurniawan change the names in setup.py file also

Comment (HandyKurniawan) [reply setup.cfg:6]: ah thank you :)

File: quconot/implementations/mct_no_auxiliary_relative.py (hunk 1 above)
Comment (HandyKurniawan) [reply quconot/implementations/mct_no_auxiliary_relative.py:16]: agree!

Comment (ankit27kh) [reply quconot/implementations/mct_no_auxiliary_relative.py:16]: #12 
created an issue so we remember this

**Comment by ankit27kh** [comment]
@HandyKurniawan do you know what's being changed in the license file?

**Comment by HandyKurniawan** [comment]
> @HandyKurniawan do you know what's being changed in the license file?

@ankit27kh I didn't change anything, I think it was from before, maybe we can use the one from the documentation branch.

File: test/implementations/functions_testing.py [hunk 7]
```diff
@@ -8,436 +7,659 @@
...
+                atol=absolute_error_tol,
+                rtol=relative_error_tol,
+            )
+            is False
+        ):
+            return False, "Something wrong with the implementation"
 
-    return True, ""
+        return True, ""
+    except Exception:
```
Comment (adamglos92) [inline test/implementations/functions_testing.py:43]: this cannot be like this. you should check what kind of exception is raised and react accordingly (actually why do you need try-except in the first place?)

File: test/implementations/functions_testing.py [hunk 8]
```diff
@@ -8,436 +7,659 @@
...
+            return False, "Something wrong with the implementation"
+
+        return True, ""
+    except Exception:
+        return False, ""
+
+
+def verify_dependencies(rd):
+    if rd["DNW"]:
+        assert rd["CNW"]
```
Comment (adamglos92) [inline test/implementations/functions_testing.py:636]: I would say this should run recursively `verify_dependencies` on `CNW`, etc.

File: test/implementations/test_mct_barenco_74_dirty.py [hunk 9]
```diff
@@ -41,4 +126,132 @@ def test_circuit_dirty_relative_auxiliary(self):
...
+
+            res, msg = verify_circuit_dirty_wasted_relative_separable_auxiliary(
+                unitary_matrix, controls_no, auxiliaries_no
+            )
+
+            self._result_dict["DWRS"] = res
+
             assert res, msg
+
+    def test_dependencies(self):
```
Comment (adamglos92) [inline test/implementations/test_mct_barenco_74_dirty.py:224]: isn't this repeated?

**Review by adamglos92 — CHANGES_REQUESTED** [review]

File: quconot/implementations/mct_no_auxiliary_relative.py [hunk 10]
```diff
@@ -1,79 +1,70 @@
...
+        :rtype: QuantumCircuit
+        """
+        qc = QuantumCircuit(self._n + 1)
+        if self._n == 2:
+            qc.rccx(0, 1, 2)
+        elif self._n == 3:
+            qc.rcccx(0, 1, 2, 3)
+
+        # should be done for all implementations
+        # TODO: solve issue with reordered qubits
```
[the remark this answers is not in the document]
Comment (adamglos92) [reply quconot/implementations/mct_no_auxiliary_relative.py:60]: after you do this resolve this conversation

File: test/implementations/functions_testing.py (hunk 8 above)
Comment (adamglos92) [reply test/implementations/functions_testing.py:636]: this is optional for this PR. But could be next thing to do

File: test/implementations/test_mct_barenco_74_dirty.py (hunk 9 above)
Comment (HandyKurniawan) [reply test/implementations/test_mct_barenco_74_dirty.py:224]: this one is for every implementation

File: test/implementations/functions_testing.py (hunk 8 above)
Comment (HandyKurniawan) [reply test/implementations/functions_testing.py:636]: i was thinking to put in the functions_testing, but it seems this one is not working. I forgot to remove this

Comment (HandyKurniawan) [reply test/implementations/functions_testing.py:636]: i will add for recursive as an issue

File: quconot/implementations/mct_no_auxiliary_relative.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (HandyKurniawan) [reply quconot/implementations/mct_no_auxiliary_relative.py:60]: This remark is from the MCT Base that you gave to use before.. I am not sure yet what to do here.

File: setup.cfg [hunk 11]
```diff
@@ -1,44 +1,44 @@
...
-exclude = test
-
-[coverage:run]
-relative_files = True
+[metadata]
+name = QuCoNot
+description = QuCoNot is a Python library for all your MCT needs.
+long_description = file: README.md
+long_description_content_type = text/markdown
+author = Adam Glos, Ankit Khandelwal, Handy Kurniawan, Shraddha Aangiras, Özlem Salehi
```
[the remark this answers is not in the document]
Comment (HandyKurniawan) [reply setup.cfg:6]: I thought Ö would be at the end of the alphabet. Should I put Shraddha at the end?

File: quconot/implementations/mct_barenco_74_dirty.py [hunk 12]
```diff
@@ -110,7 +110,7 @@ def verify_mct_cases(
         relative_phase: true / false (D)
```
[the remark this answers is not in the document]
Comment (ankit27kh) [reply quconot/implementations/mct_barenco_74_dirty.py:110]: Yes. docstrings are being handled in the documentation PR.

File: setup.py [hunk 13]
```diff
@@ -25,7 +25,7 @@
 info = {
     "name": "QuCoNot",
     "version": VERSION,
-    "author": "Adam Glos",
+    "author": "Adam Glos, Ankit Khandelwal, Handy Kurniawan, Shraddha Aangiras, Özlem Salehi",
```
[the remark this answers is not in the document]
Comment (HandyKurniawan) [reply setup.py:28]: `Shradda Aangiras, Adam Glos, Ankit Khandelwal, Handy Kurniawan, Özlem Salehi`

File: quconot/implementations/mct_barenco_74_dirty.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (HandyKurniawan) [reply quconot/implementations/mct_barenco_74_dirty.py:110]: moved to issue #13

File: quconot/implementations/mct_no_auxiliary_relative.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (HandyKurniawan) [reply quconot/implementations/mct_no_auxiliary_relative.py:60]: moved to issue #14

File: setup.cfg [hunk 14]
```diff
@@ -1,44 +1,44 @@
...
+author_email = aglos@iitis.pl
```
[the remark this answers is not in the document]
Comment (HandyKurniawan) [reply setup.cfg]: has fixed with the latest pushed

[the remark this answers is not in the document]
Comment (HandyKurniawan) [reply setup.cfg]: fixed

File: test/implementations/functions_testing.py (hunk 8 above)
Comment (HandyKurniawan) [reply test/implementations/functions_testing.py:636]: moved to issue #15

File: test/implementations/functions_testing.py (hunk 7 above)
Comment (HandyKurniawan) [reply test/implementations/functions_testing.py:43]: moved to issue #16

File: setup.cfg (hunk 11 above)
[the remark this answers is not in the document]
Comment (HandyKurniawan) [reply setup.cfg]: fixed `Shraddha Aangiras, Adam Glos, Ankit Khandelwal, Handy Kurniawan, Özlem Salehi`

File: setup.cfg (hunk 14 above)
[the remark this answers is not in the document]
Comment (HandyKurniawan) [reply setup.cfg]: `Shraddha Aangiras, Adam Glos, Ankit Khandelwal, Handy Kurniawan, Özlem Salehi`

File: test/implementations/test_mct_barenco_74_dirty.py (hunk 9 above)
Comment (HandyKurniawan) [reply test/implementations/test_mct_barenco_74_dirty.py:224]: moved to issue #15

**Review by adamglos92 — APPROVED** [review]

**Pull request merged**